tensorflow_encrypted/tensor/odd_implicit.py

```diff
@@ -0,0 +1,137 @@
+        self.value = value
+        self.dtype = tf.int32
+
+    @staticmethod
+    def from_native(value: Union[np.ndarray, tf.Tensor], dtype=tf.int32) -> 'OddImplicitTensor':
+        assert isinstance(value, (np.ndarray, tf.Tensor)), type(value)
+        return OddImplicitTensor(value, dtype=dtype)
+
+    @staticmethod
+    def sample_uniform(shape: Union[Tuple[int, ...], tf.TensorShape], dtype=tf.int32) -> 'OddImplicitTensor':
+        return OddImplicitTensor(tf.random_uniform(shape=shape, dtype=dtype,
+                                                   minval=dtype.min, maxval=dtype.max), dtype)
+
+    @staticmethod
+    def sample_bounded(shape: List[int], bitlength: int) -> 'OddImplicitTensor':
+        raise NotImplementedError()
+
+    @staticmethod
+    def stack(x: List['OddImplicitTensor'], axis: int = 0) -> 'OddImplicitTensor':
+        assert all(isinstance(i, OddImplicitTensor) for i in x)
+        return OddImplicitTensor.from_native(tf.stack([v.value for v in x], axis=axis), dtype=x[0].dtype)
+
+    @staticmethod
+    def concat(x: List['OddImplicitTensor'], axis: int) -> 'OddImplicitTensor':
+        assert all(isinstance(i, OddImplicitTensor) for i in x)
+        return OddImplicitTensor.from_native(tf.concat([v.value for v in x], axis=axis), dtype=x[0].dtype)
+
+    def eval(self, sess: tf.Session, feed_dict: Dict[Any, Any]={},
+             tag: Optional[str]=None) -> 'OddImplicitTensor':
+        return OddImplicitTensor(run(sess, self.value, feed_dict=feed_dict, tag=tag), dtype=self.dtype)
+
+    def __getitem__(self, slice: Any) -> Union[tf.Tensor, np.ndarray]:
+        return self.value[slice]
+
+    def __repr__(self) -> str:
+        return 'OddImplicitTensor(shape={}, dtype={})'.format(self.shape, self.dtype)
+
+    @property
+    def shape(self) -> Union[Tuple[int, ...], tf.TensorShape]:
+        return self.value.shape
+
+    def __add__(self, other: Union['OddImplicitTensor', int]) -> 'OddImplicitTensor':
+        return self.add(other)
+
+    def __sub__(self, other: Union['OddImplicitTensor', int]) -> 'OddImplicitTensor':
+        return self.sub(other)
+
+    def __mul__(self, other: Union['OddImplicitTensor', int]) -> 'OddImplicitTensor':
+        return self.mul(other)
+
+    def __mod__(self, k: int) -> 'OddImplicitTensor':
+        return self.mod(k)
+
+    def add(self, other: Union['OddImplicitTensor', int]) -> 'OddImplicitTensor':
+        x, y = _lift(self, self.dtype), _lift(other, self.dtype)
+
+        def func(vals):
+            x, y = vals
+
+            negative = lambda: tf.cond(x < tf.int32.min - y, lambda: x + y - 1, lambda: x + y)
+
+            maybe_negative = lambda: tf.cond(y < 0, negative, lambda: x + y)
+
+            positive = lambda: tf.cond(x > tf.int32.max - y, lambda: x + y + 1, lambda: x + y)
+
+            return tf.cond(y > 0, positive, maybe_negative)
+
+        ret = tf.map_fn(func, (x.value, y.value), dtype=self.dtype)
```

Comment:
fyi you could alternatively express this using `tf.where` (not sure which approach is fastest):
```
z = x + y

# correct for overflow where needed
z = z + tf.where(
    tf.logical_and(y > 0, x > tf.int32.max - y),
    tf.ones(z.shape, dtype=tf.int32),
    tf.zeros(z.shape, dtype=tf.int32)
)

# correct for underflow where needed
z = z - tf.where(
    tf.logical_and(y < 0, x < tf.int32.min - y),
    tf.ones(z.shape, dtype=tf.int32),
    tf.zeros(z.shape, dtype=tf.int32)
)
```